src/lib/rules/amp-validator/amp-validator.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * @fileoverview Validates the HTML of a page is AMP valid
+ */
+
+import * as amphtmlValidator from 'amphtml-validator';
+
+// The list of types depends on the events you want to capture.
+import { Category } from '../../enums/category';
+import { RuleContext } from '../../rule-context';
+import { IRule, IRuleBuilder } from '../../types';
+import { debug as d } from '../../utils/debug';
+
+const debug: debug.IDebugger = d(__filename);
+
+import { ITargetFetchEnd, IScanEnd } from '../../types';
+
+// ------------------------------------------------------------------------------
+// Public
+// ------------------------------------------------------------------------------
+
+const rule: IRuleBuilder = {
+    create(context: RuleContext): IRule {
+        let validPromise;
+        const errorsOnly = context.ruleOptions && context.ruleOptions['errors-only'] || false;
+        let html;
+
+        const onTargetFetchEnd = (fetchEnd: ITargetFetchEnd) => {
+            const { response: { body: { content }, statusCode } } = fetchEnd;
+
+            if (statusCode !== 200 || !content) {
+                return;
+            }
+
```

Comment:
@molant This rule don't make sense for non-HTML documents, so you should [check if it's a HTML document](https://github.com/sonarwhal/sonar/blob/ebb51bb8b61cecc6a04f3ed80f30ecb7599e7c6a/src/lib/utils/misc.ts#L55-L76) or not.

Reply:
In my mind everything that comes via the `targetfetch::end` should be `HTML` or at least a string. It doesn't make too much sense, at least for now. Once we start working on the online editor I think that `fetch::end` should be the way to go for the other type of files.
If we agree that anything could come in this event we should do the validation only once somewhere and not in each rule.

Reply:
> If we agree that anything could come in this event 

My opinion is that should be the case. Users should be allow to provide sonar with any URL, and sonar should handle is appropriately.

I think having this kinda of behavior contributes to a better user experience (same as we don't require users to type the `http(s)://` part and sonar will figure it out).

> we should do the validation only once somewhere and not in each rule.

Can you open an issue with your proposal so that we can discuss it? Thanks!

Reply:
> My opinion is that should be the case. Users should be allow to provide sonar with any URL, and sonar should handle is appropriately.

Maybe, but in all the documentation we point to websites when using sonar and not individual files. We never talk about analyzing an image for example. IIRC (some?) rules assume that static assets like css, js, and images come in fetch::end and html in targetfetch::end. Connectors definitely assume that, and there isn't a single test for the case you point out.

I agree we should clarify the behavior sooner rather than later, and I'll open an issue for that, but I don't think that should block merging this PR.

Reply:
> I agree we should clarify the behavior sooner rather than later, and I'll open an issue for that, but I don't think that should block merging this PR.

:+1:

Reply:
Issue opened: https://github.com/sonarwhal/sonar/issues/550